### PR TITLE
minor: fix TOKEN_TEXT_PATTERN in JavadocMetadataScraper

### LINKED ISSUE #9140: fix TOKEN_TEXT_PATTERN in JavadocMetadataScraper
`Pattern TOKEN_TEXT_PATTERN = Pattern.compile("([A-Z]+_*+)+[A-Z]+");`

detected at
https://github.com/checkstyle/checkstyle/pull/9128#discussion_r548900637 

> Why do we have * on _? Do we have any token with 2 or more underscores?
Shouldn't this be ([A-Z]+_)+[A-Z]+ ?
If there is a case with no underscore, then it should be something like ([A-Z]+_)+?[A-Z]+ .

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #9140 : fix TOKEN_TEXT_PATTERN in JavadocMetadataScraper

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java [hunk 1]
```diff
@@ -78,7 +78,7 @@
             Pattern.compile("\\s*Violation Message Keys:\\s*");
 
     /** Regular expression for detecting ANTLR tokens(for e.g. CLASS_DEF). */
-    private static final Pattern TOKEN_TEXT_PATTERN = Pattern.compile("([A-Z]+_*+)+[A-Z]+");
+    private static final Pattern TOKEN_TEXT_PATTERN = Pattern.compile("([A-Z]+_?)+[A-Z]+");
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:81]: We have patterns with no underscores at all? Can you give an example?

**Comment by rnveach** [comment]
This should be tied to the issue regarding it. https://github.com/checkstyle/checkstyle/issues/9140

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java (hunk 1 above)
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:81]: `TYPECAST`

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:81]: https://checkstyle.sourceforge.io/config_annotation.html#AnnotationOnSameLine You can find one such token here.

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:81]: `TYPE` `PARAMETERS` `EXPR`

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:81]: `RCURLY` `LCURLY`

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:81]: ah ok.

Then shouldn't the `?` be transferred to the `+` outside the `()`s since the `[A-Z]+` on the end covers it?
Example: `([A-Z]+_)*[A-Z]+`

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java]: Good point, done.

**Review by rnveach — APPROVED** [review]

**Comment by romani** [comment]
@Vyom-Yadav ,

please review CI failure, CI should be green.
it complains on fact that after your fix there are bunch of changes in files
```
✔ ~/java/github/romani/checkstyle [Vyom-Yadav/Token_Text_Pattern L|✚ 192] 
19:16 $ git diff
diff --git a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/ArrayTypeStyleCheck.xml b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/ArrayTypeStyleCheck.xml
index 82e64e5..244bfc0 100644
--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/ArrayTypeStyleCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/ArrayTypeStyleCheck.xml
@@ -4,19 +4,19 @@
       <check fully-qualified-name="com.puppycrawl.tools.checkstyle.checks.ArrayTypeStyleCheck"
              name="ArrayTypeStyle"
              parent="com.puppycrawl.tools.checkstyle.TreeWalker">
-         <description>&lt;p&gt;
+         <description>&lt;p&gt;P_TAG_START
  Checks the style of array type definitions.
- Some like Java style: {@code public static void main(String[] args)}
- and some like C style: {@code public static void main(String args[])}.
- &lt;/p&gt;
- &lt;p&gt;
+ Some like Java style: {@code public static void main(String[] args)}JAVADOC_INLINE_TAG
+ and some like C style: {@code public static void main(String args[])}JAVADOC_INLINE_TAG.
+ &lt;/p&gt;P_TAG_END
+ &lt;p&gt;P_TAG_START
  By default the Check enforces Java style.
- &lt;/p&gt;
- &lt;p&gt;
+ &lt;/p&gt;P_TAG_END
+ &lt;p&gt;P_TAG_START
  This check strictly enforces only Java style for method return types regardless
- of the value for 'javaStyle'. For example, {@code byte[] getData()}.
+ of the value for 'javaStyle'. For example, {@code byte[] getData()}JAVADOC_INLINE_TAG.
  This is because C doesn't compile methods with array declarations on the name.
- &lt;/p&gt;</description>
+ &lt;/p&gt;P_TAG_END</description>
          <properties>
             <property default-value="true" name="javaStyle" type="boolean">
                <description>Control whether to enforce Java style (true) or C style (false).</description>

```

they are not expected.

**Review by romani — CHANGES_REQUESTED** [review]
see comment above

**Comment by romani** [comment]
Github, rebase

**Comment by romani** [comment]
CI should be green
https://travis-ci.org/github/checkstyle/checkstyle/jobs/753799198#L1977
There should be no diff, or at least it should valid and be part of commit

**Comment by Vyom-Yadav** [comment]
@romani The `MetadataGeneratorUtilTest` completes successfully but every it runs, it modifies some files. I can't understand why, I feel the regex is fine, I think there is some problem related to the test.

**Comment by rnveach** [comment]
@Vyom-Yadav The test modifying the files is on purpose but it is a little confusing. If the file changes during a run it is suppose to represent a change needed in it like if you modify the checks and add a new token or property. In this case, the files changing means your regular expression change is disrupting the files and we are not expecting this type of disruption in this PR.

Look at the token @romani provided compared to your regular expression and you can see that your expression won't match it when it should.

**Comment by Vyom-Yadav** [comment]
@romani Ah got it, so we got tokens with 2 `_`

**Comment by Vyom-Yadav** [comment]
@rnveach I think this one's fine `("([A-Z]+_)*[A-Z]+")` . Is it fine or more simplification is needed?

**Comment by rnveach** [comment]
It looks fine to me.

**Comment by Vyom-Yadav** [comment]
```
-            <property default-value="CLASS_DEF,INTERFACE_DEF,ENUM_DEF,ANNOTATION_DEF,ANNOTATION_FIELD_DEF,ENUM_CONSTANT_DEF,PARAMETER_DEF,VARIABLE_DEF,METHOD_DEF,CTOR_DEF,COMPACT_CTOR_DEF,RECORD_DEF"
+            <property default-value="D,T,CLASS_DEF,INTERFACE_DEF,ENUM_DEF,ANNOTATION_DEF,ANNOTATION_FIELD_DEF,ENUM_CONSTANT_DEF,PARAMETER_DEF,VARIABLE_DEF,METHOD_DEF,CTOR_DEF,COMPACT_CTOR_DEF,RECORD_DEF"
```
@rnveach Again it's failing, it's adding some extra tokens `D` and `T`.  Should I change the regex so that it doesn't match tokens without underscores?

**Comment by Vyom-Yadav** [comment]
Also if that is to be done, then why? This is the regex for detecting ANTLR Tokens and there are ANTLR tokens with no underscores like `LCURLY`, etc.

**Comment by rnveach** [comment]
>  Should I change the regex so that it doesn't match tokens without underscores?

You said earlier we had tokens without underscores, so won't that break them?
https://github.com/checkstyle/checkstyle/pull/9153#discussion_r551762851

You will need to look into what these false letters being brought in are and how we need to fix this to only bring in actual tokens.

**Comment by Vyom-Yadav** [comment]
@rnveach @romani `([A-Z]+_)*[A-Z]{2,}` this regex should work now. I will explain why I added this `{2,}` .

- First of all, we don't have any token which is shorter than 2 characters and I don't think in the future also we would have a single character token.
- Also the files were getting modified and false tokens like `D` `T` were being added, so I figured out that our regex should not allow a single character token. The initial regex `([A-Z]+_*+)+[A-Z]+` did not allow single-character tokens, both brackets had `+` quantifier to make our regex at least 2 characters long.

**Comment by Vyom-Yadav** [comment]
@rnveach @romani ping

**Comment by romani** [comment]
@Vyom-Yadav , this will cause https://sonarcloud.io/project/issues?id=org.checkstyle%3Acheckstyle&issues=AXZvXAw5a-Kxo9081VHL&open=AXZvXAw5a-Kxo9081VHL again can you make `*+`?

we run into this problem only because of https://github.com/checkstyle/checkstyle/pull/9128

**Comment by Vyom-Yadav** [comment]
@romani So you don't want the regex to backtrack, `([A-Z]+_)*+[A-Z]{2,}`, this will still run for all cases successfully.
But how will backtracking help over HERE-- 

- suppose our token is `ABCD`, it will begin by checking the first group and it will not find a match over there, so it will move to the next character set, it will find its match and no regex is after it so it will terminate successfully. 

- now our token is `ABCD_AB`, it will begin by checking the first group, it will match with `ABCD_` once and then it (first group) will try to match with `AB` but will not match, and then it (our regex i.e. `([A-Z]+_)*+[A-Z]{2,}`) will check that if `AB` matches with the next character set, it matches, so it will terminate successfully.

- Basically what I am trying to say is that possessive quantifier will not make a difference in this case, it will never backtrack even if we don't use possessive quantifier. Please correct me if I am going the wrong way.

**Comment by rnveach** [comment]
@romani I think the reason why `"([A-Z]+_)*[A-Z]+"` didn't work shows a larger underlying problem.
https://github.com/checkstyle/checkstyle/pull/9153#issuecomment-759235102
> it's adding some extra tokens D and T.

There are no tokens `D` and `T` so the fact that it brought them in is very strange and made me looking into them more.

https://github.com/checkstyle/checkstyle/blob/1e3fb3fae940bf7ea340592f7f0cda3af1b320e3/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java#L325
`initialText` is receiving more than just the tokens. It is receiving the entire Javadoc line including other things besides the token.
> `Default value is:\n <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_TRY">\n LITERAL_TRY</a>,\n <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_CATCH">\n LITERAL_CATCH</a>,\n <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_FINALLY">\n LITERAL_FINALLY</a>,\n <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_IF">\n LITERAL_IF</a>,\n <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_ELSE">\n LITERAL_ELSE</a>.\n </li>

This explains where the `D` and `T` are coming from. D is from the word `Default` and `T` is from the word `TokenTypes` in our URL. All of which shouldn't be matching as tokens.

The fact that we are parsing an entire javadoc text and we don't know where our tokens are seems like a concern for me since this leaves us to the possibility that we can match any uppercase letters by mistake just because someone wrote something like that into the javadoc. This explains a simple regexp change to fail like this. The way things are coded right now we could accidentally match any consecutive uppercase letters, the most common example is any acronym like JRE and JDK.

This code should be looked into and possibly redone. Blind text parsing like this shows inexperience and will only lead us to more troubles in the future, IMO.

**Comment by Vyom-Yadav** [comment]
@rnveach What to do next?

**Comment by rnveach** [comment]
I would ask @romani for guidance.

To me, this code looks to be trouble and should be examined on a better way to resolve this. I don't see us finding a magic value for the regular expression to simplify things. The issue might have to be expanded.

**Comment by romani** [comment]
this become too complicated
what about https://github.com/checkstyle/checkstyle/pull/9206 ?

**Comment by Vyom-Yadav** [comment]
@rnveach @romani Shall I close it?

**Comment by romani** [comment]
fixed in https://github.com/checkstyle/checkstyle/pull/9206

@Vyom-Yadav , more precise issue is opened https://github.com/checkstyle/checkstyle/issues/9218 so you can reuse your knowledge over code to fix deeper problem, sorry that we have not clear issue on what is required.
